flexprep/domain/processing.py

```diff
@@ -98,36 +79,67 @@ def _load_and_validate_data(
-            )
-            ds_in |= metadata.extract_pv(ds_in["u"].message)
-        for temp_file in temp_files:
-            os.unlink(temp_file)
-        return ds_in
+        try:
+            with config.set_values(data_scope="ifs"):
+                source = data_source.FileDataSource(datafiles=temp_files)
+                ds_in = grib_decoder.load(source, request)
+                validate_dataset(
+                    ds_in,
+                    request["param"],
+                    to_process["forecast_ref_time"],
+                    int(to_process["step"]),
+                    int(prev_file["step"]),
+                )
+                ds_in |= metadata.extract_pv(ds_in["u"].message)
+
+            return ds_in
+
+        except Exception as e:
+            logging.error(f"Data loading and validation failed: {e}")
+            return None
+
+        finally:
+            # Ensure temporary files are cleaned up
+            for temp_file in temp_files:
+                os.unlink(temp_file)
 
     def _apply_flexpart(self, ds_in: typing.Any) -> typing.Any:
-        """Process data and return processed data structure."""
+        """Apply flexpart pre-processing and return processed data structure."""
         ds_out = flx.fflexpart(ds_in)
         prepare_output(ds_out, ds_in, INPUT_FIELDS, CONSTANTS)
         return ds_out
 
     def _save_output(
-        self, ds_out: typing.Any, forecast_ref_time: str, step_to_process: int
+        self,
+        ds_out: typing.Any,
+        forecast_ref_time: dt,
+        step_to_process: int,
+        key_to_process: str,
     ) -> None:
-        """Save processed data to a temporary file."""
-        output_file = tempfile.NamedTemporaryFile(
-            suffix=f"output_dispf{forecast_ref_time}_{step_to_process}", delete=False
-        )
-        with open(output_file.name, "wb") as fout:
-            for name, field in ds_out.items():
-                if field.attrs.get("v_coord") == "hybrid":
-                    logging.info(f"Writing GRIB fields to {output_file.name}")
-                    grib_decoder.save(field, fout)
+        """Save processed data to a temporary file and upload to output-S3."""
+        forecast_ref_time_str = forecast_ref_time.strftime("%Y%m%d%H%M")
+
+        try:
+            key = f"output_dispf{forecast_ref_time_str}_step{step_to_process}"
+
+            with tempfile.NamedTemporaryFile(
+                suffix=key,
+                delete=False,
+            ) as output_file:
+                # Write data to the temporary file
+                with open(output_file.name, "wb") as fout:
+                    for name, field in ds_out.items():
+                        if field.attrs.get("v_coord") == "hybrid":
+                            logging.info(f"Writing GRIB fields to {output_file.name}")
+                            grib_decoder.save(field, fout)
+
+            # Upload the file to S3
+            S3client().upload_file(output_file.name, key=key)
```

Comment:
If this line is moved into the context manager above, the `delete=False` can be removed. Otherwise, you will need to remove the tempfile at some point